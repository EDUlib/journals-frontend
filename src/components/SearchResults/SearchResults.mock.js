import {
  FILTER_ID_ALL, FILTER_ID_DOCUMENTS, FILTER_ID_IMAGES, FILTER_ID_VIDEOS,
} from '../../data/constants/filterTypes';

const mockSearchResults = {
  meta: {
    total_count: 3,
    text_count: 0,
    image_count: 1,
    video_count: 1,
    doc_count: 1,
  },
  hits: [
    {
      block_id: 15,
      block_title: 'block 1',
      block_type: 'pdf',
      breadcrumbs: ['x', 'y'],
      highlights: [
        '1 <b>foo </b>',
        '2 <b>foo </b>',
        '3 <b>foo </b>',
        '4 <b>foo </b>',
        '5 <b>foo </b>',
      ],
      journal_about_page_id: 12,
      journal_id: 5,
      journal_name: 'new journal',
      page_id: 15,
      page_title: 'pdf document foo',
      score: '2.1442184',
      span_id: '',
    },
    {
      page_id: 13,
      page_title: 'video',
      breadcrumbs: [],
      journal_about_page_id: 12,
      journal_id: 5,
      journal_name: 'new journal',
      block_id: 2,
      block_title: 'Science and Cooking Chef Profile: JOSÉ ANDRÉS',
      block_type: 'xblock_video',
      highlights: [
        'video-title: performance director <b>foo</b>',
      ],
      score: '1.2321665',
      span_id: 'xblock_video-c81e728d9d4c2f636f067f89cc14862c',
    },
    {
      page_id: 13,
      page_title: 'a page',
      breadcrumbs: [],
      journal_about_page_id: 12,
      journal_id: 5,
      journal_name: 'new journal',
      block_id: 1,
      block_title: 'image-title: foo',
      block_type: 'image',
      highlights: [
        'image-title: <b>foo</b>',
      ],
      score: '12.110976',
      span_id: 'image-c4ca4238a0b923820dcc509a6f75849b',
    },
  ],
};

const mockGetResultCount = jest.fn((filter) => {
  switch (filter) {
    case FILTER_ID_ALL:
      return mockSearchResults.hits.total_count;
    case FILTER_ID_DOCUMENTS:
      return mockSearchResults.hits.doc_count;
    case FILTER_ID_IMAGES:
      return mockSearchResults.hits.image_count;
    case FILTER_ID_VIDEOS:
      return mockSearchResults.hits.video_count;
    default:
      return 0;
  }
});

export { mockGetResultCount, mockSearchResults };

