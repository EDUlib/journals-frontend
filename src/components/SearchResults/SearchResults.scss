.results-container {
  margin-top: 20px;
}

.card-link {
  a:hover {
      text-decoration: none;
  }

  .search-result-card {
    display: block;
    padding: 10px 30px 0px 30px;
    text-align: left;
    border-radius: 4px;
    background-color: $white;
    border: solid 1px #d6dbe0;

    &:hover{
        background-color: $pale-grey;
        cursor: pointer;
    }
  }
}

.result-icon {
  font-size: 21px;
  line-height: 1.67;
  color: $gray-700;
}

.result-title {
  padding-left: 8px;
  font-family: "Open Sans";
  font-size: 20px;
  font-weight: 600;
  color: $deep-sea-blue;
  text-decoration: underline;
}

.result-highlight {
  font-family: "Open Sans";
  font-size: 16px;
  color: $dark-grey;
  display: block;
  padding-bottom: 10px;

  b {
    font-weight: bold;
    background-color:#dceef7;
    padding: 0px 4px 0px 4px;
    border-radius: 4px;
  }
}

.additional-highlights {
  font-family: "Open Sans";
  font-size: 16px;
  letter-spacing: normal;
  color: $slate-grey;
  padding-bottom: 10px;
}

.breadcrumb div{
  display: inline-block;
}

.result-breadcrumb {
  display: inline-block;
  font-family: "Open Sans";
  font-size: 14px;
  font-weight: 600;
  line-height: 1.71;
  letter-spacing: 1px;
  color: $slate-grey;
  padding-right: 4px;
  text-transform: uppercase
}

.result-breadcrumb:not(:last-child):after{
  content: ':';
}

.result-progress {
  margin: 20px;
  text-align: left;
}
