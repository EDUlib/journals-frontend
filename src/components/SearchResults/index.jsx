import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { HashLink as Link } from 'react-router-hash-link';
import { Icon } from '@edx/paragon';
import { FormattedPlural } from 'react-intl';

import './SearchResults.scss';
import {
  RICH_TEXT,
  RAW_HTML,
  PDF,
  XBLOCK_VIDEO,
  IMAGE,
} from '../../data/constants/contentTypes';

import {
  FILTER_ID_ALL,
  FILTER_ID_IMAGES,
  FILTER_ID_DOCUMENTS,
  FILTER_ID_VIDEOS,
} from '../../data/constants/filterTypes';

const NUM_HIGHLIGHTS_DISPLAY = 3;

const BreadCrumb = ({
  breadcrumbs,
}) => (
  <div className="breadcrumb">
    {
    breadcrumbs.map(breadcrumb => (
      <div key={breadcrumb} className="result-breadcrumb">
        {breadcrumb}
      </div>
    ))
  }
  </div>
);

const AdditionalHighlights = ({
  total,
}) => (
  total > NUM_HIGHLIGHTS_DISPLAY &&
    <div className="additional-highlights">
      <FormattedPlural
        value={total - NUM_HIGHLIGHTS_DISPLAY}
        one={`+ ${total - NUM_HIGHLIGHTS_DISPLAY} more instance`}
        other={`+ ${total - NUM_HIGHLIGHTS_DISPLAY} more instances`}
      />
    </div>
);

const SearchHighlight = ({
  highlights,
}) => (
  highlights.map(highlight => (
    <div
      key={highlight}
      className="result-highlight"
      dangerouslySetInnerHTML={{ __html: highlight }} // eslint-disable-line react/no-danger
    />
  ))
);

const SearchItem = ({
  data,
}) => (
  <Link className="card-link" to={`/${data.journal_about_page_id}/pages/${data.page_id}#${data.span_id}`} >
    <div className="search-result-card">
      <div className="result-icon">
        <Icon
          className={
            classNames({
              fa: true,
              'fa-book': data.block_type === RICH_TEXT || data.block_type === RAW_HTML,
              'fa-file-video-o': data.block_type === XBLOCK_VIDEO,
              'fa-picture-o': data.block_type === IMAGE,
              'fa-file-pdf-o': data.block_type === PDF,
            }).split(' ')
          }
        />
        <span className="result-title">{data.block_title}</span>
      </div>

      <SearchHighlight highlights={data.highlights.slice(0, NUM_HIGHLIGHTS_DISPLAY)} />
      <AdditionalHighlights total={data.highlights.length} />
      <BreadCrumb breadcrumbs={data.breadcrumbs} />
    </div>
  </Link>
);

class SearchResults extends React.Component {
  formatKey(searchResults) {
    return `${searchResults.journal_id}-${searchResults.page_id}-${searchResults.block_id}-${searchResults.block_type}`;
  }

  matchesFilter(blockType) {
    return (this.props.filter === FILTER_ID_ALL ||
            (this.props.filter === FILTER_ID_DOCUMENTS && blockType === PDF) ||
              (this.props.filter === FILTER_ID_IMAGES && blockType === IMAGE) ||
              (this.props.filter === FILTER_ID_VIDEOS && blockType === XBLOCK_VIDEO));
  }

  render() {
    if (this.props.error) {
      return (
        <div className="result-progress"> Error occurred searching</div>
      );
    }
    if (this.props.searchStarted && !this.props.searchFinished) {
      return (
        <div className="result-progress"> Searching...</div>
      );
    }
    if (this.props.resultsCounter(this.props.filter) === 0) {
      return (
        <div className="result-progress">No matching results found</div>
      );
    }
    return (
      <div>
        <div className="results-container">
          {
            this.props.hits.map(searchResult => (
              this.matchesFilter(searchResult.block_type) &&
                <SearchItem
                  key={this.formatKey(searchResult)}
                  data={searchResult}
                  filter={this.props.filter}
                />
            ))
          }
        </div>
      </div>
    );
  }
}

BreadCrumb.defaultProps = {
  breadcrumbs: [],
};

BreadCrumb.propTypes = {
  breadcrumbs: PropTypes.arrayOf(PropTypes.string),
};

SearchItem.defaultProps = {
};

SearchItem.propTypes = {
  data: PropTypes.shape({
    journal_about_page_id: PropTypes.number,
    page_id: PropTypes.number,
    span_id: PropTypes.string,
    block_title: PropTypes.string,
    block_type: PropTypes.string,
    highlights: PropTypes.array,
    breadcrumbs: PropTypes.array,
  }).isRequired,
};

SearchResults.defaultProps = {
  hits: [],
  searchFinished: true,
  searchStarted: false,
  error: null,
  resultsCounter: () => 0,
};

SearchResults.propTypes = {
  resultsCounter: PropTypes.func,
  filter: PropTypes.string.isRequired,
  searchFinished: PropTypes.bool,
  searchStarted: PropTypes.bool,
  error: PropTypes.instanceOf(Error),
  hits: PropTypes.arrayOf(PropTypes.object),
};

export default SearchResults;
