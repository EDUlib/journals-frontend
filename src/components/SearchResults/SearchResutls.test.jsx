import React from 'react';
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';
import SearchResults from '.';
import {
  FILTER_ID_ALL, FILTER_ID_DOCUMENTS, FILTER_ID_IMAGES, FILTER_ID_VIDEOS,
} from '../../data/constants/filterTypes';


import { mockGetResultCount, mockSearchResults } from './SearchResults.mock';

const SearchResultWrapper = props => (
  <IntlProvider locale="en">
    <MemoryRouter>
      <SearchResults
        {...props}
      />
    </MemoryRouter>
  </IntlProvider>
);

describe('<SearchResults/>', () => {
  describe('renders', () => {
    it('with no searchResults', () => {
      const tree = renderer
        .create((
          <SearchResultWrapper filter={FILTER_ID_ALL} />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with filter all', () => {
      const tree = renderer
        .create((
          <SearchResultWrapper
            resultsCounter={mockGetResultCount}
            filter={FILTER_ID_ALL}
            hits={mockSearchResults.hits}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with filter document', () => {
      const tree = renderer
        .create((
          <SearchResultWrapper
            resultsCounter={mockGetResultCount}
            filter={FILTER_ID_DOCUMENTS}
            hits={mockSearchResults.hits}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with filter image ', () => {
      const tree = renderer
        .create((
          <SearchResultWrapper
            resultsCounter={mockGetResultCount}
            filter={FILTER_ID_IMAGES}
            hits={mockSearchResults.hits}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with filter video ', () => {
      const tree = renderer
        .create((
          <SearchResultWrapper
            resultsCounter={mockGetResultCount}
            filter={FILTER_ID_VIDEOS}
            hits={mockSearchResults.hits}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with wrong filter', () => {
      const tree = renderer
        .create((
          <SearchResultWrapper
            filter="foobar"
            resultsCounter={mockGetResultCount}
            hits={mockSearchResults.hits}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with error', () => {
      const tree = renderer
        .create((
          <SearchResultWrapper
            filter={FILTER_ID_ALL}
            hits={mockSearchResults.hits}
            error={new Error('some error')}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with no finish loading', () => {
      const tree = renderer
        .create((
          <SearchResultWrapper
            filter={FILTER_ID_VIDEOS}
            hits={mockSearchResults.hits}
            searchStarted
            searchFinished={false}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
