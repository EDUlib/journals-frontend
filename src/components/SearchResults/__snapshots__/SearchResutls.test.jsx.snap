// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResults/> renders with error 1`] = `
<div
  className="result-progress"
>
   Error occurred searching
</div>
`;

exports[`<SearchResults/> renders with filter all 1`] = `
<div>
  <div
    className="results-container"
  >
    <a
      className="card-link"
      href="/12/pages/15#"
      onClick={[Function]}
    >
      <div
        className="search-result-card"
      >
        <div
          className="result-icon"
        >
          <span
            aria-hidden={true}
            className="fa fa-file-pdf-o"
            id="Icon1"
          />
          <span
            className="result-title"
          >
            block 1
          </span>
        </div>
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "1 <b>foo </b>",
            }
          }
        />
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "2 <b>foo </b>",
            }
          }
        />
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "3 <b>foo </b>",
            }
          }
        />
        <div
          className="additional-highlights"
        >
          <span>
            + 2 more instances
          </span>
        </div>
        <div
          className="breadcrumb"
        >
          <div
            className="result-breadcrumb"
          >
            new journal
          </div>
          <div
            className="result-breadcrumb"
          >
            x
          </div>
          <div
            className="result-breadcrumb"
          >
            y
          </div>
        </div>
      </div>
    </a>
    <a
      className="card-link"
      href="/12/pages/13#xblock_video-c81e728d9d4c2f636f067f89cc14862c"
      onClick={[Function]}
    >
      <div
        className="search-result-card"
      >
        <div
          className="result-icon"
        >
          <span
            aria-hidden={true}
            className="fa fa-file-video-o"
            id="Icon1"
          />
          <span
            className="result-title"
          >
            Science and Cooking Chef Profile: JOSÉ ANDRÉS
          </span>
        </div>
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "video-title: performance director <b>foo</b>",
            }
          }
        />
        <div
          className="breadcrumb"
        >
          <div
            className="result-breadcrumb"
          >
            new journal
          </div>
        </div>
      </div>
    </a>
    <a
      className="card-link"
      href="/12/pages/13#image-c4ca4238a0b923820dcc509a6f75849b"
      onClick={[Function]}
    >
      <div
        className="search-result-card"
      >
        <div
          className="result-icon"
        >
          <span
            aria-hidden={true}
            className="fa fa-picture-o"
            id="Icon1"
          />
          <span
            className="result-title"
          >
            image-title: foo
          </span>
        </div>
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "image-title: <b>foo</b>",
            }
          }
        />
        <div
          className="breadcrumb"
        >
          <div
            className="result-breadcrumb"
          >
            new journal
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;

exports[`<SearchResults/> renders with filter document 1`] = `
<div>
  <div
    className="results-container"
  >
    <a
      className="card-link"
      href="/12/pages/15#"
      onClick={[Function]}
    >
      <div
        className="search-result-card"
      >
        <div
          className="result-icon"
        >
          <span
            aria-hidden={true}
            className="fa fa-file-pdf-o"
            id="Icon1"
          />
          <span
            className="result-title"
          >
            block 1
          </span>
        </div>
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "1 <b>foo </b>",
            }
          }
        />
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "2 <b>foo </b>",
            }
          }
        />
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "3 <b>foo </b>",
            }
          }
        />
        <div
          className="additional-highlights"
        >
          <span>
            + 2 more instances
          </span>
        </div>
        <div
          className="breadcrumb"
        >
          <div
            className="result-breadcrumb"
          >
            new journal
          </div>
          <div
            className="result-breadcrumb"
          >
            x
          </div>
          <div
            className="result-breadcrumb"
          >
            y
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;

exports[`<SearchResults/> renders with filter image  1`] = `
<div>
  <div
    className="results-container"
  >
    <a
      className="card-link"
      href="/12/pages/13#image-c4ca4238a0b923820dcc509a6f75849b"
      onClick={[Function]}
    >
      <div
        className="search-result-card"
      >
        <div
          className="result-icon"
        >
          <span
            aria-hidden={true}
            className="fa fa-picture-o"
            id="Icon1"
          />
          <span
            className="result-title"
          >
            image-title: foo
          </span>
        </div>
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "image-title: <b>foo</b>",
            }
          }
        />
        <div
          className="breadcrumb"
        >
          <div
            className="result-breadcrumb"
          >
            new journal
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;

exports[`<SearchResults/> renders with filter video  1`] = `
<div>
  <div
    className="results-container"
  >
    <a
      className="card-link"
      href="/12/pages/13#xblock_video-c81e728d9d4c2f636f067f89cc14862c"
      onClick={[Function]}
    >
      <div
        className="search-result-card"
      >
        <div
          className="result-icon"
        >
          <span
            aria-hidden={true}
            className="fa fa-file-video-o"
            id="Icon1"
          />
          <span
            className="result-title"
          >
            Science and Cooking Chef Profile: JOSÉ ANDRÉS
          </span>
        </div>
        <div
          className="result-highlight"
          dangerouslySetInnerHTML={
            Object {
              "__html": "video-title: performance director <b>foo</b>",
            }
          }
        />
        <div
          className="breadcrumb"
        >
          <div
            className="result-breadcrumb"
          >
            new journal
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;

exports[`<SearchResults/> renders with no finish loading 1`] = `
<div
  className="result-progress"
>
   Searching...
</div>
`;

exports[`<SearchResults/> renders with no searchResults 1`] = `
<div
  className="result-progress"
>
  No matching results found
</div>
`;

exports[`<SearchResults/> renders with wrong filter 1`] = `
<div
  className="result-progress"
>
  No matching results found
</div>
`;
